### PR TITLE
Reusable blocks: Use view context, always include title.raw and content.raw

### LINKED ISSUE #11343: Fetch reusable blocks in the view context on first render.
**Describe the bug**
When inserting, the editor fetches reusable blocks in via the API in the `view` context. This permits authors and contributors to insert blocks created by other users.

On first render, the editor fetches reusable blocks via the REST API in the `edit` context. When an author goes back to edit a post that uses another user's block, the block renders with the error "Block has been deleted or is unavailable"

Originally logged in the discussion of [WP#45098 (comment)](https://core.trac.wordpress.org/ticket/45098#comment:24).

**To Reproduce**
1. Create two users, `user1` (any role) and `user2` (author role)
1. Log in as `user1` and create a reusable block
1. Log in as `user2` and create a post using `user1`'s block
1. The block renders correctly.
1. Save the post and refresh the edit screen.
1. The block fails to render and displays the error "Block has been deleted or is unavailable." The REST API returns a 403 error.

It looks to be happening around these lines but when I change the context to view, the REST API returns a 200 but the blocks still fail to render.

https://github.com/WordPress/gutenberg/blob/b7d8fbfd0b97d579f732d9fc8f026e2919c2642d/packages/editor/src/store/effects/reusable-blocks.js#L65-L70

**Expected behavior**
Reusable blocks render in the view context when first loading the edit screen.

**Screenshots**
Screencast of author creating a post that re-uses a block created by an admin. **This is recorded against WordPress 5.0**

author

**Desktop (please complete the following information):**
 - OS: macOS
 - Browser Firefox
 - Version 63

### PR DESCRIPTION
**Pull request by noisysocks**
Fixes #11343.

Existing reusable blocks on a post would not render if the user was an author or a contributor. This happened because requests to fetch a single block or post are blocked when ?context=edit is passed and the current user is not an editor.

The fix is to stop passing `?context=edit` to block requests, but this causes another problem which is that Gutenberg needs access to both `title.raw` and `content.raw` in the API response.

We therefore both stop passing `?context=edit` to block requests *and* modify the blocks API so that `title.raw` and `content.raw` are always provided.

#### How to test

1. Create two users, `user1` (any role) and `user2` (author role)
1. Log in as `user1` and create a reusable block
1. Log in as `user2` and create a post using `user1`'s block
1. The block should render correctly.
1. Save the post and refresh the edit screen.
1. The block should again render correctly.

### CODE REVIEW

**Comment by noisysocks** [comment]
Noting that this will require a commit to the 5.0 branch which I can take care of once approved.

File: lib/class-wp-rest-blocks-controller.php [hunk 1]
```diff
@@ -33,4 +33,21 @@ public function check_read_permission( $post ) {
...
+	 * Retrieves the block's schema, conforming to JSON Schema.
+	 *
+	 * @since 4.4.0
+	 *
+	 * @return array Item schema data.
+	 */
+	public function get_item_schema() {
+		$schema = parent::get_item_schema();
+
+		// Allow all contexts to access the raw title and content of a block.
```
Comment (tofumatt) [inline lib/class-wp-rest-blocks-controller.php]: Doesn't this… do exactly the opposite? We're `unset`ing the variable. Maybe there is some PHP nuance I'm missing out on 😅

File: phpunit/class-rest-blocks-controller-test.php [hunk 2]
```diff
@@ -101,7 +101,7 @@ public function test_capabilities( $action, $role, $expected_status ) {
 				$request->set_body_params(
 					array(
 						'title'   => 'Test',
-						'content' => '<!-- wp:core/paragraph --><p>Test</p><!-- /wp:core/paragraph -->',
+						'content' => '<!-- wp:paragraph --><p>Test</p><!-- /wp:paragraph -->',
```
Comment (tofumatt) [inline phpunit/class-rest-blocks-controller-test.php]: Are these test changes related to the schema change? I wouldn't expect the block name to change here.

**Review by tofumatt — COMMENTED** [review]
I'm… a bit unclear on why `unset`ing variables causes them to be included in the schema. Maybe I'm lost.

**Comment by peterwilsoncc** [comment]
@noisysocks I'll take a look this weekend and test the patch against core too.

Copying in @kadamwhite as he may have opinions too.

File: lib/class-wp-rest-blocks-controller.php (hunk 1 above)
Comment (noisysocks) [reply lib/class-wp-rest-blocks-controller.php]: The reason that `content.raw` doesn't appear in all contexts is because, when declaring the `content.raw` property, the schema specifies `'context' => array( 'edit' )`:

https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2016

`unset()`ing this `context` value means that the property will appear in all contexts.

File: phpunit/class-rest-blocks-controller-test.php (hunk 2 above)
Comment (noisysocks) [reply phpunit/class-rest-blocks-controller-test.php]: No, not related. I just noticed that the test data here was out of date (we stopped prefixing block markup with `core/` a while ago) 🙂

Comment (tofumatt) [reply phpunit/class-rest-blocks-controller-test.php]: Huh, it's odd that the test didn't fail before then...

**Comment by noisysocks** [comment]
Closing as this isn't quite right. Working with @peterwilsoncc on an alternative approach.

File: phpunit/class-rest-blocks-controller-test.php (hunk 2 above)
Comment (mtias) [reply phpunit/class-rest-blocks-controller-test.php]: We still support the core prefix to not break older blocks.

Comment (tofumatt) [reply phpunit/class-rest-blocks-controller-test.php]: Ah, maybe we should have a test for both then? 😉
